IdentityCore/src/cache/mac/MSIDMacAppCredentialCacheItem.m

```diff
@@ -0,0 +1,194 @@
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#import "MSIDMacAppCredentialCacheItem.h"
+#import "NSDictionary+MSIDExtensions.h"
+
+static NSString *keyDelimiter = @"-";
+
+@interface MSIDMacAppCredentialCacheItem ()
+
+@property (nonatomic) NSMutableDictionary *cacheObjects;
+@property (nonatomic) dispatch_queue_t queue;
+
+@end
+
+@implementation MSIDMacAppCredentialCacheItem
+
+- (instancetype)initPrivate
+{
+    if(self = [super init])
+    {
+        self.cacheObjects = [NSMutableDictionary dictionary];
+        self.queue = dispatch_queue_create("com.microsoft.universalStorage", DISPATCH_QUEUE_CONCURRENT);
+    }
+    
+    return self;
+}
+
++ (MSIDMacAppCredentialCacheItem *)sharedInstance
```

Comment:
why it needs to be a singleton? Can there be multiple blobs one app is dealing with?

Reply:
Singleton is important here since there can be multiple concurrent threads reading or writing to this dictionary, we may expose the creation of the dictionary to multiple threads thus ending up in a situation where one thread creates the object and starts using it, while another one is creating a new instance at the same time because when it last checked, the instance didn’t exist